### ISSUE TITLE
GSL Fails to Compile with Exceptions Disabled

### ISSUE THREAD

**Issue by aheysse** [microsoft/GSL#635]
gsl::narrow will fail to compile on clang with exceptions disabled. It doesn't seem to honor 

GSL_UNENFORCED_ON_CONTRACT_VIOLATION or
GSL_THROW_ON_CONTRACT_VIOLATION or
GSL_UNENFORCED_ON_CONTRACT_VIOLATION,

but rather just blindly throws an exception. I assume GSL_CONTRACT_CHECK is meant to be used.

For good measure, here's the compiler error:
1>D:\Github\handheld\src-external\gsl\include\gsl/gsl_util(109,33): error : cannot use 'throw' with exceptions disabled
1>    if (static_cast<U>(t) != u) throw narrowing_error();
1>                                ^
1>D:\Github\handheld\src-external\gsl\include\gsl/gsl_util(111,9): error : cannot use 'throw' with exceptions disabled
1>        throw narrowing_error();
1>        ^

**Comment by aheysse** [comment]
@neilmacintosh Thanks for the quick update. Our team is in the process of taking GSL as a submodule so I look forward to the fix. Thx!

**Comment by annagrin** [comment]
Note that _HAS_EXCEPTIONS=0 in MS STL is not an official story for no exception mode. Current resolution of the issue is experimental only, until we have an official support for no-exception mode in the language and STL.

**Comment by JordanMaples** [comment]
Maintainer's call: It looks like this issue has already been handled, however I'm closing the issue as the contract violation behavior has further changed to only result in termination.

**Comment by phillipjohnston** [comment]
Hi @JordanMaples,

Is this actually closed? I still get this error with GSL 3.1.0:

```
In file included from ../subprojects/GSL-3.1.0/include/gsl/gsl:23:
../subprojects/GSL-3.1.0/include/gsl/gsl_util:110:33: error: cannot use 'throw' with exceptions disabled
    if (static_cast<U>(t) != u) throw narrowing_error{};
                                ^
../subprojects/GSL-3.1.0/include/gsl/gsl_util:112:9: error: cannot use 'throw' with exceptions disabled
        throw narrowing_error{};
        ^
```

Looking at the narrow function, it doesn't seem like there's any condition that causes the `throw` to be avoided.

```
// narrow() : a checked version of narrow_cast() that throws if the cast changed the value
template <class T, class U>
GSL_SUPPRESS(type.1) // NO-FORMAT: attribute
GSL_SUPPRESS(f.6) // NO-FORMAT: attribute // TODO: MSVC /analyze does not recognise noexcept(false)
constexpr
T narrow(U u) noexcept(false)
{
    T t = narrow_cast<T>(u);
    if (static_cast<U>(t) != u) throw narrowing_error{};
    if (!details::is_same_signedness<T, U>::value && ((t < T{}) != (u < U{})))
        throw narrowing_error{};
    return t;
}
```

**Comment by phillipjohnston** [comment]
I see that the change was reverted in commit 61534ca3adda0e214d6fb7dd9bc4173edb6e02d1. I'll open a new issue.
